File: internal/service/student_service.go
package service

import (
	"errors"
	"teaching-management/internal/model"
	"teaching-management/internal/repository"

	"github.com/google/uuid"
)

type StudentService struct {
	studentRepo *repository.StudentRepository
}

func NewStudentService(studentRepo *repository.StudentRepository) *StudentService {
	return &StudentService{studentRepo: studentRepo}
}

func (s *StudentService) CreateStudent(userID uuid.UUID, req *model.StudentRequest) (*model.StudentResponse, error) {
	if req.FullName == "" {
		return nil, errors.New("student full name is required")
	}
	if req.FeeModel != "per_session" && req.FeeModel != "monthly" && req.FeeModel != "per_hour" {
		return nil, errors.New("invalid fee model; must be per_session, monthly, or per_hour")
	}

	student := &model.Student{
		UserID:    userID,
		FullName:  req.FullName,
		BirthDate: req.BirthDate,
		Gender:    req.Gender,
		Address:   req.Address,
		School:    req.School,
		Grade:     req.Grade,
		Phone:     req.Phone,
		Notes:     req.Notes,
		FeeModel:  req.FeeModel,
		FeeAmount: req.FeeAmount,
	}

	parent := &model.Parent{
		FatherName: req.FatherName,
		MotherName: req.MotherName,
		Phones:     req.Phones,
		Email:      req.Email,
		Address:    req.ParentAddr,
		Occupation: req.Occupation,
	}

	err := s.studentRepo.Create(student, parent, req.Subjects)
	if err != nil {
		return nil, err
	}

	return s.studentRepo.GetByID(student.ID, userID)
}

func (s *StudentService) GetStudents(userID uuid.UUID) ([]model.StudentResponse, error) {
	return s.studentRepo.GetAll(userID)
}

func (s *StudentService) GetStudentByID(userID uuid.UUID, id uuid.UUID) (*model.StudentResponse, error) {
	res, err := s.studentRepo.GetByID(id, userID)
	if err != nil {
		return nil, err
	}
	if res == nil {
		return nil, errors.New("student not found")
	}
	return res, nil
}

func (s *StudentService) UpdateStudent(userID uuid.UUID, id uuid.UUID, req *model.StudentRequest) error {
	existing, err := s.studentRepo.GetByID(id, userID)
	if err != nil {
		return err
	}
	if existing == nil {
		return errors.New("student not found")
	}

	if req.FullName == "" {
		return errors.New("student full name is required")
	}
	if req.FeeModel != "per_session" && req.FeeModel != "monthly" && req.FeeModel != "per_hour" {
		return errors.New("invalid fee model")
	}

	student := &model.Student{
		ID:        id,
		UserID:    userID,
		FullName:  req.FullName,
		BirthDate: req.BirthDate,
		Gender:    req.Gender,
		Address:   req.Address,
		School:    req.School,
		Grade:     req.Grade,
		Phone:     req.Phone,
		Notes:     req.Notes,
		FeeModel:  req.FeeModel,
		FeeAmount: req.FeeAmount,
	}

	parent := &model.Parent{
		StudentID:  id,
		FatherName: req.FatherName,
		MotherName: req.MotherName,
		Phones:     req.Phones,
		Email:      req.Email,
		Address:    req.ParentAddr,
		Occupation: req.Occupation,
	}

	return s.studentRepo.Update(student, parent, req.Subjects)
}

func (s *StudentService) DeleteStudent(userID uuid.UUID, id uuid.UUID) error {
	existing, err := s.studentRepo.GetByID(id, userID)
	if err != nil {
		return err
	}
	if existing == nil {
		return errors.New("student not found")
	}

	return s.studentRepo.Delete(id, userID)
}

File: src/services/student.service.ts
import { StudentRepository } from '../repositories/interfaces.js';
import { StudentRequest, StudentResponse, Parent } from '../types/index.js';

export class StudentService {
  constructor(private studentRepo: StudentRepository) {}

  async createStudent(userId: string, req: StudentRequest): Promise<StudentResponse> {
    if (!req.full_name || req.full_name.trim() === '') {
      throw new Error('student full name is required');
    }

    if (
      req.fee_model !== 'per_session' &&
      req.fee_model !== 'monthly' &&
      req.fee_model !== 'per_hour'
    ) {
      throw new Error('invalid fee model; must be per_session, monthly, or per_hour');
    }

    const studentData = {
      user_id: userId,
      full_name: req.full_name.trim(),
      birth_date: req.birth_date || null,
      gender: req.gender || '',
      address: req.address || '',
      school: req.school || '',
      grade: req.grade || '',
      phone: req.phone || '',
      notes: req.notes || '',
      fee_model: req.fee_model,
      fee_amount: Number(req.fee_amount || 0)
    };

    const parentData: Parent = {
      father_name: req.father_name || '',
      mother_name: req.mother_name || '',
      phones: req.phones || [],
      email: req.email || '',
      address: req.parent_address || '',
      occupation: req.occupation || ''
    };

    return this.studentRepo.create(studentData, parentData, req.subjects || []);
  }

  async getStudents(userId: string): Promise<StudentResponse[]> {
    return this.studentRepo.getAll(userId);
  }

  async getStudentById(userId: string, id: string): Promise<StudentResponse> {
    const res = await this.studentRepo.getById(id, userId);
    if (!res) {
      throw new Error('student not found');
    }
    return res;
  }

  async updateStudent(userId: string, id: string, req: StudentRequest): Promise<void> {
    const existing = await this.studentRepo.getById(id, userId);
    if (!existing) {
      throw new Error('student not found');
    }

    if (!req.full_name || req.full_name.trim() === '') {
      throw new Error('student full name is required');
    }

    if (
      req.fee_model !== 'per_session' &&
      req.fee_model !== 'monthly' &&
      req.fee_model !== 'per_hour'
    ) {
      throw new Error('invalid fee model');
    }

    const studentData = {
      id,
      user_id: userId,
      full_name: req.full_name.trim(),
      birth_date: req.birth_date || null,
      gender: req.gender || '',
      address: req.address || '',
      school: req.school || '',
      grade: req.grade || '',
      phone: req.phone || '',
      notes: req.notes || '',
      fee_model: req.fee_model,
      fee_amount: Number(req.fee_amount || 0)
    };

    const parentData: Parent = {
      student_id: id,
      father_name: req.father_name || '',
      mother_name: req.mother_name || '',
      phones: req.phones || [],
      email: req.email || '',
      address: req.parent_address || '',
      occupation: req.occupation || ''
    };

    await this.studentRepo.update(studentData, parentData, req.subjects);
  }

  async deleteStudent(userId: string, id: string): Promise<void> {
    const existing = await this.studentRepo.getById(id, userId);
    if (!existing) {
      throw new Error('student not found');
    }
    await this.studentRepo.delete(id, userId);
  }
}
